### PR TITLE
try to reconect when rtc.connect returns undefined

### PR DESCRIPTION
**Pull request by alenjdev**
when creating a real-time connection `device.startRealTimeConnecton()` we proceed even if the peerID was not returned, so we get in this idle state that is not connected but does not show any error. 

The proposal it's to keep trying to get the id every two seconds, not sure if there is a case when we will never get an id back, which will cause an endless loop.

### CODE REVIEW

File: packages/data-sdk/src/Device.ts [hunk 1]
```diff
@@ -274,7 +274,13 @@ export class Device extends EventEmitter implements IRealtimeDevice {
 
       // We can connect our real-time communication client to device peers by their ID
       this.remoteDevicePeerId = devicePeer.id;
-      await (rtcClient as RtcClient).connect(this.remoteDevicePeerId);
+      let isConnected = null;
+      while (typeof isConnected !== "string") {
+        isConnected = await (rtcClient as RtcClient).connect(
+          this.remoteDevicePeerId
```
Comment (ericbarch) [inline packages/data-sdk/src/Device.ts:280]: Seems like this would cause an infinite loop if the device ever goes offline and its peer ID expires. Since `this.remoteDevicePeerId` is updated on line 276, it will never be refreshed once in this loop.

Maybe every X failed attempts, it would make sense to execute this logic. Could just break it out as a separate method. https://github.com/FormantIO/toolkit/pull/78/files#diff-f777f3fac8c979f45751b42876eea470219441adcb722899e6022fd7ca85cf73R265-R276

**Review by ericbarch — CHANGES_REQUESTED** [review]

File: packages/data-sdk/src/Device.ts (hunk 1 above)
Comment (alenjdev) [reply packages/data-sdk/src/Device.ts:280]: it does cause an infinite loop, I'll update it to try a few times. and just for some context, this issue happens when we try to start a `observe` session, when a teleop session is connected already, it fails like 50% of the time.

File: packages/data-sdk/demo/main.ts [hunk 2]
```diff
@@ -1,20 +1,26 @@
...
-        console.log(d);
-      });
+      const d = await Fleet.getCurrentDevice();
+      let conected = false;
+      while (!conected) {
+        console.warn("waiting for main connection");
+        conected = await d.isInRealtimeSession();
+        await timeout(2000);
+      }
+      console.warn("star onnection");
```
Comment (ericbarch) [inline packages/data-sdk/demo/main.ts:21]: Nit; typo

File: packages/data-sdk/package.json [hunk 3]
```diff
@@ -26,7 +26,7 @@
   },
   "version": "0.0.147",
   "scripts": {
-    "dev": "vite --port 9146",
+    "dev": "vite --port 5173",
```
Comment (ericbarch) [inline packages/data-sdk/package.json:29]: Hmm, why the change in ports?

File: packages/data-sdk/src/Device.ts [hunk 4]
```diff
@@ -262,38 +263,63 @@ export class Device extends EventEmitter implements IRealtimeDevice {
...
   }
 
+  private async getRemoteDevicePeerId(rtcClient: RtcClient | RtcClientV1) {
+    // Each online device and user has a peer in the system
+    const peers = await rtcClient.getPeers();
+
+    // Find the device peer corresponding to the device's ID
+    const devicePeer = peers.find((_) => _.deviceId === this.id);
+
+    if (!isRtcPeer(devicePeer)) {
```
Comment (ericbarch) [inline packages/data-sdk/src/Device.ts:300]: Out of curiosity, why do you need to do this vs just checking if the peer object exists?

File: packages/data-sdk/src/Device.ts [hunk 5]
```diff
@@ -262,38 +263,63 @@ export class Device extends EventEmitter implements IRealtimeDevice {
...
+      throw new Error("Cannot find peer, is the robot offline?");
+    }
+    return devicePeer.id;
+  }
+
+  private async createSession(rtcClient: RtcClient | RtcClientV1) {
+    // We can connect our real-time communication client to device peers by their ID
+    const tries = 3;
+    if (this.remoteDevicePeerId) {
+      for (let i = 0; i < tries; i++) {
```
Comment (ericbarch) [inline packages/data-sdk/src/Device.ts:311]: Would it make sense to have a delay in this loop?

File: packages/data-sdk/src/Device.ts [hunk 6]
```diff
@@ -262,38 +263,63 @@ export class Device extends EventEmitter implements IRealtimeDevice {
...
       // Wait for the signaling process to complete...
-      while (
-        rtcClient.getConnectionStatus(this.remoteDevicePeerId) !== "connected"
-      ) {
-        await delay(100);
-      }
-      this.rtcClient = rtcClient;
-      this.emit("connect");
+      if (!!sessionId) {
+        while (
```
Comment (ericbarch) [inline packages/data-sdk/src/Device.ts:274]: Wouldn't it still be possible to get stuck in this loop? I.e. maybe the internet suddenly drops and now this RTC session will never connect.

File: packages/data-sdk/package.json (hunk 3 above)
Comment (alenjdev) [reply packages/data-sdk/package.json:29]: I have so many custom modules/data-sdk/ui-sdk that I have to add to different orgs to test them out, and it is just a headache to go and manually >settings>modules>URL>url schema. It is just easy to have them in the same port so I can just add them to a view or to teleop and test.

File: packages/data-sdk/src/Device.ts (hunk 4 above)
Comment (alenjdev) [reply packages/data-sdk/src/Device.ts:300]: I hit that error a few times yesterday, but now that I rethink it there is a possibility that I was actually not getting the `devicePeer`. I think I did print the object and it was there but still hit the error.

File: packages/data-sdk/src/Device.ts [hunk 7]
```diff
@@ -262,38 +270,72 @@ export class Device extends EventEmitter implements IRealtimeDevice {
...
-      }
-      this.rtcClient = rtcClient;
-      this.emit("connect");
 
-      this.initConnectionMonitoring();
+      if (!!sessionId) {
+        const tries = 5;
+        for (let i = 0; i < tries; i++) {
+          const connectionCompleted =
+            rtcClient.getConnectionStatus(this.remoteDevicePeerId) !==
```
Comment (ericbarch) [inline packages/data-sdk/src/Device.ts:285]: Inverted logic?

**Review by ericbarch — CHANGES_REQUESTED** [review]

**Review by ericbarch — APPROVED** [review]
